README
NAME
    Business::DK::Phonenumber - Danish telephone number validator/formatter

VERSION
    This documentation describes version 0.01

SYNOPSIS
        use Business::DK::Phonenumber qw(validate render);
    
        #Validation
        if (Business::DK::Phonenumber->validate($phonenumber)) { ... }
    
        #Default format
        print Business::DK::Phonenumber->render($phonenum);
        # +45 12 34 56 78
    
        #Brief human readable Danish phone number format
        print Business::DK::Phonenumber->render($phonenum, '%08d');
        # 12345678
    
        #Normal human readable Danish phonenumber format
        print Business::DK::Phonenumber->render($phonenum, '%02d %02d %02d %02d');
        # 12 34 56 78

        #Generation of a single random number
        Business::DK::Phonenumber->generate();
    
        #Generation of 100 random numbers, using another template
        Business::DK::Phonenumber->generate(100, '%02d %02d %02d %02d');

DESCRIPTION
    This module offers functionality to validate, format and generate Danish
    phone numbers.

    The validation can recognise telephone numbers is the following formats
    as Danish phone numbers.

    *   12345678

    *   4512345678

    *   +4512345678

    White space characters are ignored.

    In addition to validation the module offers generation of valid danish
    phone numbers. The purpose of using generated phone number is up to the
    user, but the original intent is generation of varied sets of test data.

    If you want to use OOP please have a look at:

    *   Class::Business::DK::Phonenumber

    If you are using Data::FormValidator

    *   Data::FormValidator::Constraints::Business::DK::Phonenumber

SUBROUTINES AND METHODS
    The following subroutines are to be used in a procedural manner.

  validate($phonenumber)
    This subroutine takes a string and validated whether it is a Danish
    phone number.

    Returns 1 (true) or 0 (false), depending on validity.

  render($phonenumber, $template)
    Returns a phonenumber rendered according to the template parameter or
    the default.

  generate($template, $amount)
    This subroutine takes a string representing a phone number template and
    generates the amount specified by second argument: amount. If no amount
    is specified only a single random phone number is returned.

    The subroutine returns an array, no matter what amount is specified.

  _generate($template)
    This is the actual generating method used by "generate".

    It takes a single parameter, a string indicating the template for the
    formatting of the phone numbers to be generated.

    It returns a single random number representing a Danish phone number
    formatted as outlined by the specified template.

  validate_template
    This method is used internally to validate template parameters. Please
    refer to Perl's sprintf for documentation.

DIAGNOSTICS
    *   phone number not in recognisable format, the phone number provided
        to the constructor is not parsable. Please evaluate what you are
        attempting to feed to the constructor.

    *   phone number parameter is mandatory for the constructor, please
        specify the phone number parameter to the constructor in order to
        continue.

    *   template not in recognisable format, the template provided to the
        constructor is not in a parsable format, please evaluate what you
        attempting to feed to the constructor.

CONFIGURATION AND ENVIRONMENT
    No special configuration or environment is necessary.

DEPENDENCIES
    *   Carp

    *   Exporter

INCOMPATIBILITIES
    No known incompatibilities at this time.

BUGS AND LIMITATIONS
    No known bugs or limitations at this time.

TEST AND QUALITY
    *   The
        Perl::Critic::Policy::ValuesAndExpressions::RequireNumberSeparators
        policy has been disabled. We are working with phonenumbers, strings
        consisting primarily of number, so not special interpretation or
        calculative behaviour is needed.

    *   Perl::Critic::Policy::ValuesAndExpressions::ProhibitConstantPragma
        policy has been disabled. I like constants.

    *   Perl::Critic::Policy::InputOutput::RequireBracedFileHandleWithPrint
        policy has been disabled for now should be revisited at some point.

TODO
    *   Please refer to the distribution TODO file

SEE ALSO
    sprintf
        Business::DK::Phonenumber utilizes sprintf to as templating system
        for formatting telephonenumbers. This is a well specified and tested
        interface which is easy to use.

    Class::Business::DK::Phonenumber
    Data::FormValidator::Constraints::Business::DK::Phonenumber

BUG REPORTING
    Please report issues via CPAN RT:

    http://rt.cpan.org/NoAuth/Bugs.html?Dist=Business-DK-Phonenumber

    or by sending mail to

    bug-Business-DK-Phonenumber@rt.cpan.org

MOTIVATION
    I have been working in Telco for a long time. So validation and
    formatting of telephone numbers is something I have seen at lot of. This
    module is an attempt to sort of consolidate the numerous different
    regular expression solutions I have seen scathered over large code
    bases.

AUTHOR
    Jonas B. Nielsen, (jonasbn) - "<jonasbn@cpan.org>"

COPYRIGHT
    Business-DK-Phonenumber is (C) by Jonas B. Nielsen, (jonasbn) 2008-2009

LICENSE
    Business-DK-Phonenumber is released under the artistic license

    The distribution is licensed under the Artistic License, as specified by
    the Artistic file in the standard perl distribution
    (http://www.perl.com/language/misc/Artistic.html).



Changes
Changes file for Business::DK::Phonenumber

$Id: Changes 8198 2013-08-02 09:33:21Z jonasbn $

Project site changelog:

http://logicLAB.jira.com/browse/BDKPHN#selectedTab=com.atlassian.jira.plugin.system.project%3Achangelog-panel

0.06 2013-08-02

- Fixed up Changes file as part of my Questhub quest adhering to the standard
  described in: CPAN::Changes::Spec [BDKPHN-12]

  Quest: http://questhub.io/realm/perl/quest/51f0fd8e18ba7dcf4b000040
  Spec: https://metacpan.org/module/CPAN::Changes::Spec
  
  https://logiclab.jira.com/browse/BDKPHN-12

- Addressed failing Kwalitee test [BDKPHN-6]

  https://logiclab.jira.com/browse/BDKPHN-6

- Added contemporay boiler plate of Perl::Critic test [BDKPHN-8]  
  
  https://logiclab.jira.com/browse/BDKPHN-8

- Prerequisites test now under TEST_AUTHOR flag [BDKPHN-5]  
  
  https://logiclab.jira.com/browse/BDKPHN-5

- Changed creation of traditional Makefile.PL to Build.PL [BDKPHN-11]  
  
  https://logiclab.jira.com/browse/BDKPHN-11


0.05 2012-04-11

- Bug fix release, update not required

- Addressed failing Kwalitee test [BDKPHN-6]


0.04 2012-01-12

- Update recommended

- fixed bug in render subroutine, template parameter was not respected
  [BDKPHN-10]


0.03 2009-07-17

- Update recommended

- fixed bug in primary regular expression in validation subroutine [BDKPHN-4]


0.02 2009-07-01

- - Maintenance release, update not required

- Failing tests for release 0.01, missing requirement: Data::FormValidator
  [BDKPHN-2]
    
- Failing tests for release 0.01, missing requirement: Test::Exception
  [BDKPHN-3]


0.01 2009-06-19

- Initial release


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.4005
use ExtUtils::MakeMaker;
WriteMakefile
(
  'NAME' => 'Business::DK::Phonenumber',
  'VERSION_FROM' => 'lib/Business/DK/Phonenumber.pm',
  'PREREQ_PM' => {
                   'Carp' => '0',
                   'Data::FormValidator' => 0,
                   'Exporter' => '0',
                   'Taint::Runtime' => 0,
                   'Test::Exception' => 0,
                   'Test::Kwalitee' => 0,
                   'Test::More' => 0,
                   'Test::Perl::Critic' => 0,
                   'Test::Prereq::Build' => 0,
                   'Test::Taint' => 0
                 },
  'INSTALLDIRS' => 'site',
  'EXE_FILES' => [],
  'PL_FILES' => {},
  'test' => {
              'TESTS' => 't/*.t t/.svn/*.t t/.svn/prop-base/*.t t/.svn/props/*.t t/.svn/text-base/*.t t/.svn/tmp/*.t t/.svn/tmp/prop-base/*.t t/.svn/tmp/props/*.t t/.svn/tmp/text-base/*.t t/Business/*.t t/Business/.svn/*.t t/Business/.svn/prop-base/*.t t/Business/.svn/props/*.t t/Business/.svn/text-base/*.t t/Business/.svn/tmp/*.t t/Business/.svn/tmp/prop-base/*.t t/Business/.svn/tmp/props/*.t t/Business/.svn/tmp/text-base/*.t t/Business/DK/*.t t/Business/DK/.svn/*.t t/Business/DK/.svn/prop-base/*.t t/Business/DK/.svn/props/*.t t/Business/DK/.svn/text-base/*.t t/Business/DK/.svn/tmp/*.t t/Business/DK/.svn/tmp/prop-base/*.t t/Business/DK/.svn/tmp/props/*.t t/Business/DK/.svn/tmp/text-base/*.t t/Business/DK/Phonenumber/*.t t/Business/DK/Phonenumber/.svn/*.t t/Business/DK/Phonenumber/.svn/prop-base/*.t t/Business/DK/Phonenumber/.svn/props/*.t t/Business/DK/Phonenumber/.svn/text-base/*.t t/Business/DK/Phonenumber/.svn/tmp/*.t t/Business/DK/Phonenumber/.svn/tmp/prop-base/*.t t/Business/DK/Phonenumber/.svn/tmp/props/*.t t/Business/DK/Phonenumber/.svn/tmp/text-base/*.t t/Class/*.t t/Class/.svn/*.t t/Class/.svn/prop-base/*.t t/Class/.svn/props/*.t t/Class/.svn/text-base/*.t t/Class/.svn/tmp/*.t t/Class/.svn/tmp/prop-base/*.t t/Class/.svn/tmp/props/*.t t/Class/.svn/tmp/text-base/*.t t/Class/Business/*.t t/Class/Business/.svn/*.t t/Class/Business/.svn/prop-base/*.t t/Class/Business/.svn/props/*.t t/Class/Business/.svn/text-base/*.t t/Class/Business/.svn/tmp/*.t t/Class/Business/.svn/tmp/prop-base/*.t t/Class/Business/.svn/tmp/props/*.t t/Class/Business/.svn/tmp/text-base/*.t t/Class/Business/DK/*.t t/Class/Business/DK/.svn/*.t t/Class/Business/DK/.svn/prop-base/*.t t/Class/Business/DK/.svn/props/*.t t/Class/Business/DK/.svn/text-base/*.t t/Class/Business/DK/.svn/tmp/*.t t/Class/Business/DK/.svn/tmp/prop-base/*.t t/Class/Business/DK/.svn/tmp/props/*.t t/Class/Business/DK/.svn/tmp/text-base/*.t t/Class/Business/DK/Phonenumber/*.t t/Class/Business/DK/Phonenumber/.svn/*.t t/Class/Business/DK/Phonenumber/.svn/prop-base/*.t t/Class/Business/DK/Phonenumber/.svn/props/*.t t/Class/Business/DK/Phonenumber/.svn/text-base/*.t t/Class/Business/DK/Phonenumber/.svn/tmp/*.t t/Class/Business/DK/Phonenumber/.svn/tmp/prop-base/*.t t/Class/Business/DK/Phonenumber/.svn/tmp/props/*.t t/Class/Business/DK/Phonenumber/.svn/tmp/text-base/*.t t/Data/*.t t/Data/.svn/*.t t/Data/.svn/prop-base/*.t t/Data/.svn/props/*.t t/Data/.svn/text-base/*.t t/Data/.svn/tmp/*.t t/Data/.svn/tmp/prop-base/*.t t/Data/.svn/tmp/props/*.t t/Data/.svn/tmp/text-base/*.t t/Data/FormValidator/*.t t/Data/FormValidator/.svn/*.t t/Data/FormValidator/.svn/prop-base/*.t t/Data/FormValidator/.svn/props/*.t t/Data/FormValidator/.svn/text-base/*.t t/Data/FormValidator/.svn/tmp/*.t t/Data/FormValidator/.svn/tmp/prop-base/*.t t/Data/FormValidator/.svn/tmp/props/*.t t/Data/FormValidator/.svn/tmp/text-base/*.t t/Data/FormValidator/Constraints/*.t t/Data/FormValidator/Constraints/.svn/*.t t/Data/FormValidator/Constraints/.svn/prop-base/*.t t/Data/FormValidator/Constraints/.svn/props/*.t t/Data/FormValidator/Constraints/.svn/text-base/*.t t/Data/FormValidator/Constraints/.svn/tmp/*.t t/Data/FormValidator/Constraints/.svn/tmp/prop-base/*.t t/Data/FormValidator/Constraints/.svn/tmp/props/*.t t/Data/FormValidator/Constraints/.svn/tmp/text-base/*.t t/Data/FormValidator/Constraints/Business/*.t t/Data/FormValidator/Constraints/Business/.svn/*.t t/Data/FormValidator/Constraints/Business/.svn/prop-base/*.t t/Data/FormValidator/Constraints/Business/.svn/props/*.t t/Data/FormValidator/Constraints/Business/.svn/text-base/*.t t/Data/FormValidator/Constraints/Business/.svn/tmp/*.t t/Data/FormValidator/Constraints/Business/.svn/tmp/prop-base/*.t t/Data/FormValidator/Constraints/Business/.svn/tmp/props/*.t t/Data/FormValidator/Constraints/Business/.svn/tmp/text-base/*.t t/Data/FormValidator/Constraints/Business/DK/*.t t/Data/FormValidator/Constraints/Business/DK/.svn/*.t t/Data/FormValidator/Constraints/Business/DK/.svn/prop-base/*.t t/Data/FormValidator/Constraints/Business/DK/.svn/props/*.t t/Data/FormValidator/Constraints/Business/DK/.svn/text-base/*.t t/Data/FormValidator/Constraints/Business/DK/.svn/tmp/*.t t/Data/FormValidator/Constraints/Business/DK/.svn/tmp/prop-base/*.t t/Data/FormValidator/Constraints/Business/DK/.svn/tmp/props/*.t t/Data/FormValidator/Constraints/Business/DK/.svn/tmp/text-base/*.t t/Data/FormValidator/Constraints/Business/DK/Phonenumber/*.t t/Data/FormValidator/Constraints/Business/DK/Phonenumber/.svn/*.t t/Data/FormValidator/Constraints/Business/DK/Phonenumber/.svn/prop-base/*.t t/Data/FormValidator/Constraints/Business/DK/Phonenumber/.svn/props/*.t t/Data/FormValidator/Constraints/Business/DK/Phonenumber/.svn/text-base/*.t t/Data/FormValidator/Constraints/Business/DK/Phonenumber/.svn/tmp/*.t t/Data/FormValidator/Constraints/Business/DK/Phonenumber/.svn/tmp/prop-base/*.t t/Data/FormValidator/Constraints/Business/DK/Phonenumber/.svn/tmp/props/*.t t/Data/FormValidator/Constraints/Business/DK/Phonenumber/.svn/tmp/text-base/*.t'
            }
)
;


lib/Data/FormValidator/Constraints/Business/DK/Phonenumber.pm
package Data::FormValidator::Constraints::Business::DK::Phonenumber;

# $Id: Phonenumber.pm 7778 2011-09-16 20:13:42Z jonasbn $

use strict;
use warnings;
use vars qw($VERSION @EXPORT_OK);
use Business::DK::Phonenumber qw(validate render);
use base qw(Exporter);

$VERSION   = '0.01';
@EXPORT_OK = qw(valid_dk_phonenumber match_dk_phonenumber);

use constant VALID   => 1;
use constant INVALID => undef;

sub valid_dk_phonenumber {
    return sub {
        my $dfv = shift;

        $dfv->name_this('valid_dk_phonenumber');

        my $phonenumber = $dfv->get_current_constraint_value();

        if ( validate($phonenumber) ) {
            return VALID;
        } else {
            return INVALID;
        }
        }
}

sub match_dk_phonenumber {
    my ( $dfv, $format ) = @_;

    my $phonenumber = ref $dfv ? $dfv->get_current_constraint_value : $dfv;

    my ($untainted_phonenumber) = render($phonenumber);

    return $untainted_phonenumber;
}

1;

__END__

=pod

=head1 NAME

Data::FormValidator::Constraints::Business::DK::Phonenumber - a DFV constraints wrapper

=head1 VERSION

This documentation describes version 0.01

=head1 SYNOPSIS

    use Data::FormValidator::Business::DK::Phonenumber qw(valid_dk_phonenumber);

    my $dfv_profile = {
        required => [qw(phonenumber)],
        constraint_methods => {
            phonenumber => valid_dk_phonenumber(),
        }
    };

    my $input_hash;
    my $result;

    $input_hash = {
        phonenumber  => 1234567,
    };

    $result = Data::FormValidator->check($input_hash, $dfv_profile);

=head1 DESCRIPTION

This module implements experimental L<Data::FormValidator> constraints.

=head1 SUBROUTINES AND METHODS

=head2 valid_dk_phonenumber

This method validates a string for resemblance to a valid Danish telephone number.

=head2 match_dk_phonenumber

This subroutine can be used for untainting a parameter.

=head1 DIAGNOSTICS

=over

=item * Please refer to L<Data::FormValidator>

=back

=head1 CONFIGURATION AND ENVIRONMENT

The module is intended for use with L<Data::FormValidator>.

=head1 DEPENDENCIES

=over

=item * L<Business::DK::Phonenumber>

=item * L<Data::FormValidator>

=back

=head1 INCOMPATIBILITIES

No known incompatibilities at this time.

=head1 BUGS AND LIMITATIONS

No known bugs or limitations at this time.

=head1 TEST AND QUALITY

=over

=item * The L<Perl::Critic::Policy::ValuesAndExpressions::RequireNumberSeparators>
policy has been disabled. We are working with phonenumbers, strings consisting primarily of number, so not special interpretation or calculative behaviour is needed.

=item * L<Perl::Critic::Policy::ValuesAndExpressions::ProhibitConstantPragma> policy has been disabled. I like constants.

=item * L<Perl::Critic::Policy::InputOutput::RequireBracedFileHandleWithPrint> policy has been disabled for now should be revisited at some point.

=back

=head1 TODO

=over

=item * Please refer to the distribution TODO file

=back

=head1 SEE ALSO

=over

=item * L<Data::FormValidator>

=item * L<Data::FormValidator::Constraints>

=item * L<Business::DK::Phonenumber>

=item * L<Class::Business::DK::Phonenumber>

=back

=head1 BUG REPORTING

Please report issues via CPAN RT:

http://rt.cpan.org/NoAuth/Bugs.html?Dist=Business-DK-Phonenumber

or by sending mail to

bug-Business-DK-Phonenumber@rt.cpan.org

=head1 MOTIVATION

I have been working in Telco for a long time. So validation and formatting of
telephone numbers is something I have seen at lot of. This module is an attempt
to sort of consolidate the numerous different regular expression solutions
I have seen scathered over large code bases.

=head1 AUTHOR

Jonas B. Nielsen, (jonasbn) - C<< <jonasbn@cpan.org> >>

=head1 COPYRIGHT

Data-FormValidator-Business-DK-Phonenumber is (C) by Jonas B. Nielsen, (jonasbn)
2008-2009

=head1 LICENSE

Data-FormValidator-Business-DK-Phonenumber is released under the artistic license

The distribution is licensed under the Artistic License, as specified
by the Artistic file in the standard perl distribution
(http://www.perl.com/language/misc/Artistic.html).

=cut


Build.PL
# Build.PL for Business-DK-Phonenumber

# $Id: Build.PL 7705 2011-06-07 10:50:48Z jonasbn $

use Module::Build;
my $build = Module::Build->new(
    dist_author       => 'Jonas B. Nielsen, (jonasbn) <jonasbn@cpan.org>',
    module_name       => 'Business::DK::Phonenumber',
    dist_version_from => 'lib/Business/DK/Phonenumber.pm',
    dist_abstract => 'Module to validate and format Danish telephonenumbers',
    license       => 'artistic',
    create_makefile_pl   => 'traditional',
    create_readme        => 1,
    recursive_test_files => 1,
    build_requires       => {
        'Test::More'          => '0',    #core
        'Test::Taint'         => '0',
        'Taint::Runtime'      => '0',
        'Data::FormValidator' => '0',
        'Test::Exception'     => '0',
		'Test::Kwalitee'     => '0',
		'Test::Perl::Critic'     => '0',
		'Test::Prereq::Build'     => '0',
    },
    requires => {
        'Exporter' => '0',               #core
        'Carp'     => '0',               #core
    },
    meta_merge => {
        resources => {
            homepage => 'http://logicLAB.jira.com/browse/BDKPHN',
            bugtracker =>
                'http://rt.cpan.org/NoAuth/Bugs.html?Dist=Business-DK-Phonenumber',
            repository => 'http://logicLAB.jira.com/svn/BDKPHN',
        }
    },
);
$build->create_build_script();
